### PR TITLE
Add log to PSI stages

### PR DESCRIPTION
**Pull request by juanli16**
* Add logging using `github.com/go-logr/stdr` which implements `github.com/go-logr/logr` interfaces using standardlib `log` package. This provides `Info` with different verbosity levels. For example, a message that was logged with `logger.V(1).Info(msg)` would not show if the verbosity level is set to 0.
* Add flags `-v` to set the verbosity in `example/sender/main.go` and `example/receiver/main.go`
* add a `logging` subsection in README on how to pass a logger to sender and receiver.

Sample logs (dhpsi):
sender:
```bash
2021/11/11 11:16:12 "level"=1 "msg"="Starting stage 1" "protocol"="dhpsi"
2021/11/11 11:16:12 "level"=1 "msg"="Finished stage 1" "protocol"="dhpsi"
2021/11/11 11:16:12 "level"=1 "msg"="Starting stage 2" "protocol"="dhpsi"
2021/11/11 11:16:12 "level"=1 "msg"="Finished stage 2" "protocol"="dhpsi"
2021/11/11 11:16:12 "level"=1 "msg"="sender finished" "protocol"="dhpsi"
```
 
receiver
```bash
2021/11/11 11:16:08 "level"=1 "msg"="Starting stage 1" "protocol"="dhpsi"
2021/11/11 11:16:08 "level"=1 "msg"="Finished stage 1" "protocol"="dhpsi"
2021/11/11 11:16:08 "level"=1 "msg"="Starting stage 2.2" "protocol"="dhpsi"
2021/11/11 11:16:08 "level"=1 "msg"="Starting stage 2.1" "protocol"="dhpsi"
2021/11/11 11:16:08 "level"=1 "msg"="Finished stage 2.1" "protocol"="dhpsi"
2021/11/11 11:16:08 "level"=1 "msg"="Finished stage 2.2" "protocol"="dhpsi"
2021/11/11 11:16:08 "level"=1 "msg"="receiver finished" "protocol"="dhpsi" "intersected"=100
```

### CODE REVIEW

File: examples/sender/main.go [hunk 1]
```diff
@@ -57,18 +59,23 @@ func main() {
 	}
 
 	log.Printf("operating with protocol %s", *protocol)
+	// fetch stdr logger
+	slog := matchlog.GetLogger(*verbose)
+	slog.V(1).Info("Should see this")
```
Comment (xcapaldi) [inline examples/sender/main.go:64]:
```suggestion
```

File: pkg/npsi/receiver.go [hunk 2]
```diff
@@ -30,11 +31,15 @@ func NewReceiver(rw io.ReadWriter) *Receiver {
...
 func (r *Receiver) Intersect(ctx context.Context, n int64, identifiers <-chan []byte) ([][]byte, error) {
+	// fetch logger
+	var logger = log.GetLoggerFromContextWithName(ctx, "npsi")
 	var intersected [][]byte
 	var k = make([]byte, hash.SaltLength)
 
 	// stage 1: P2 samples a random salt K and sends it to P1.
 	stage1 := func() error {
+		logger.Info("Starting stage 1")
+		logger.V(1).Info("Testing this logger")
```
Comment (xcapaldi) [inline pkg/npsi/receiver.go:42]:
```suggestion
```

File: pkg/npsi/receiver.go [hunk 3]
```diff
@@ -44,12 +49,15 @@ func (r *Receiver) Intersect(ctx context.Context, n int64, identifiers <-chan []
...
 
+		logger.Info("Finished stage 1")
 		return nil
 	}
 
 	// stage 2: P2 receives hashes from P1 and computes the intersection with its own hashes
 	stage2v2 := func() error {
-		//
+		logger.Info("Starting stage 2")
+		logger.V(2).Info("Testing this logger with verbosity = 2")
```
Comment (xcapaldi) [inline pkg/npsi/receiver.go:59]:
```suggestion
```

File: pkg/npsi/sender.go [hunk 4]
```diff
@@ -30,20 +31,29 @@ func NewSender(rw io.ReadWriter) *Sender {
...
 func (s *Sender) Send(ctx context.Context, n int64, identifiers <-chan []byte) error {
+	// fetch logger
+	var logger = log.GetLoggerFromContextWithName(ctx, "npsi")
+
 	// hold k
 	var k = make([]byte, hash.SaltLength)
 	// stage 1: receive a random salt K from P1
 	stage1 := func() error {
+		logger.Info("Starting stage 1")
+		logger.V(1).Info("Testing this logger")
```
Comment (xcapaldi) [inline pkg/npsi/sender.go:42]:
```suggestion
```

File: pkg/npsi/sender.go [hunk 5]
```diff
@@ -30,20 +31,29 @@ func NewSender(rw io.ReadWriter) *Sender {
...
 		}
+
+		logger.Info("Finished stage 1")
 		return nil
 	}
 
 	// stage 2: send hashes salted with K to P1
 	stage2 := func() error {
+		logger.Info("Starting stage 2")
+		logger.V(2).Info("Testing this logger with verbosity = 2")
```
Comment (xcapaldi) [inline pkg/npsi/sender.go:56]:
```suggestion
```

**Review by Yoshiji — APPROVED** [review]

File: pkg/log/log.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
+package log
```
Comment (ratelle) [inline pkg/log/log.go:1]: I think that this entire module is unnecessary for now. We won't build a logging abstraction layer so we should simply embrace logr. I would update the README to mention that we use logr internally and give a quick example of how one should call logr.NewContext to pass us a logger. GetLogger should be part of the example CLI and we should call logr.FromContext directly from the code. No need to allocate a logger and force logging on library users. We also shouldn't change the name. To expose the protocol anme, use logr.WithValues

File: README.md [hunk 7]
```diff
@@ -22,6 +22,48 @@ The naive, [highway hash](https://github.com/google/highwayhash) based PSI: an *
...
+## logging
+
+`github.com/go-logr/logr` is used internally for logging, which accepts a `logr.Logger` object. See the documentation on `logr` for various implementation of `logr.Logger`. Example implementation of match sender and receiver uses `github.com/go-logr/stdr` which logs to `os.Stderr`.
+
+### pass logger to sender or receiver
+To pass a logger to a sender or a receiver, create a new context with the parent context and `logr.Logger` object as follows
+```golang
+// for new sender with logger
+ctx := logr.NewContext(parentCtx, logger)
+sender, err := psi.NewSender(psiType, ctx, conn)
```
Comment (ratelle) [inline README.md:34]: The doc needs to be adjusted. We pass the context to the Send function, not to NewSender. Idem for receiver.

**Review by ratelle — APPROVED** [review]

**Pull request merged**